tiny_dnn/layers/max_unpooling_layer.h

```diff
@@ -62,74 +60,70 @@ class max_unpooling_layer : public feedforward_layer<Activation> {
+  serial_size_t fan_out_size() const override { return in2out_[0].size(); }
 
-  void forward_propagation(serial_size_t index,
-                           const std::vector<vec_t *> &in_data,
-                           std::vector<vec_t *> &out_data) override {
-    const vec_t &in = *in_data[0];
-    // vec_t&       out = *out_data[0];
-    vec_t &a = *out_data[1];
-    std::vector<serial_size_t> &max_idx =
-      max_unpooling_layer_worker_storage_[index].in2outmax_;
+  void forward_propagation(const std::vector<tensor_t *> &in_data,
+                           std::vector<tensor_t *> &out_data) override {
+    for (size_t sample = 0; sample < in_data[0]->size(); sample++) {
+      const vec_t &in = (*in_data[0])[sample];
+      vec_t &out      = (*out_data[0])[sample];
 
-    for_(parallelize_, 0, in2out_.size(), [&](const blocked_range &r) {
-      for (int i = r.begin(); i < r.end(); i++) {
-        const auto &in_index = out2in_[i];
-        a[i] = (max_idx[in_index] == i) ? in[in_index] : float_t{0};
-      }
-    });
+      std::vector<serial_size_t> &max_idx =
+        max_unpooling_layer_worker_storage_[sample].in2outmax_;
 
-    this->forward_activation(*out_data[0], *out_data[1]);
+      for_(parallelize_, 0, in2out_.size(), [&](const blocked_range &r) {
+        for (int i = r.begin(); i < r.end(); i++) {
+          const auto &in_index = out2in_[i];
+          out[i] = (max_idx[in_index] == i) ? in[in_index] : float_t{0};
+        }
+      });
+    }
   }
 
-  void back_propagation(serial_size_t index,
-                        const std::vector<vec_t *> &in_data,
-                        const std::vector<vec_t *> &out_data,
-                        std::vector<vec_t *> &out_grad,
-                        std::vector<vec_t *> &in_grad) override {
-    vec_t &prev_delta = *in_grad[0];
-    vec_t &curr_delta = *out_grad[1];
-    std::vector<serial_size_t> &max_idx =
-      max_unpooling_layer_worker_storage_[index].in2outmax_;
-
-    CNN_UNREFERENCED_PARAMETER(in_data);
-
-    this->backward_activation(*out_grad[0], *out_data[0], curr_delta);
-
-    for_(parallelize_, 0, in2out_.size(), [&](const blocked_range &r) {
-      for (int i = r.begin(); i != r.end(); i++) {
-        serial_size_t outi = out2in_[i];
-        prev_delta[i] = (max_idx[outi] == i) ? curr_delta[outi] : float_t{0};
-      }
-    });
+  void back_propagation(const std::vector<tensor_t *> &in_data,
+                        const std::vector<tensor_t *> &out_data,
+                        std::vector<tensor_t *> &out_grad,
+                        std::vector<tensor_t *> &in_grad) override {
+    for (serial_size_t sample = 0; sample < in_data[0]->size(); sample++) {
+      vec_t &prev_delta = (*in_grad[0])[sample];
+      vec_t &curr_delta = (*out_grad[0])[sample];
+
+      std::vector<serial_size_t> &max_idx =
+        max_unpooling_layer_worker_storage_[sample].in2outmax_;
+
+      for_(parallelize_, 0, in2out_.size(), [&](const blocked_range &r) {
```

Comment:
@karandesai-96 I know outer loop represents the samples of a mini-batch. Good fix.

But it seems to me accessing `max_unpooling_layer_worker_storage_` member variable with `sample` index is wrong.

Talking about whether the outer loop or 2nd loop should be parallelized, I can't say much without actuall testing both of the cases.